[Merged by Bors] - feat: characterizations of `IsBigO` along `atTop`

dupuisf opened the pull request:
This PR adds a way to characterize `IsBigO` along the `atTop` filter, for cases where we want the constant to depend on a "starting point" `n₀`.

It also adds the lemma `tendsto_nhds_of_eventually_eq`.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ADedecker commented:
I have suggestions for this PR that I didn't yet have the time to write down, including some proof simplification. Could you wait for them before merging? I should be able to write them in a few hours.

ADedecker reviewed (commented):
Sorry for the delay!

ADedecker commented on Mathlib/Analysis/Asymptotics/Asymptotics.lean:
```diff
@@ -2197,6 +2197,44 @@ theorem IsLittleO.nat_cast_atTop {R : Type _} [StrictOrderedSemiring R] [Archime
+theorem isBigO_atTop_iff_dep_const' {α : Type _}
+    [SemilatticeSup α] [Nonempty α] {f : α → G} {g : α → G'} :
+    f =O[atTop] g ↔ ∀ᶠ n₀ in atTop, ∃ c > 0, ∀ n ≥ n₀, c * ‖f n‖ ≤ ‖g n‖ := by
+  refine ⟨fun h => ?mp, fun h => ?mpr⟩
+  case mp =>
+    rw [isBigO_iff''] at h
+    obtain ⟨c, ⟨hc_mem, hc⟩⟩ := h
+    filter_upwards [eventually_forall_ge_atTop.mpr hc] with x hx
+    exact ⟨c, ⟨hc_mem, hx⟩⟩
+  case mpr =>
+    rw [isBigO_iff'']
+    rw [eventually_atTop] at h
+    obtain ⟨n₀, hn₀⟩ := h
+    have hn₀ := hn₀ n₀ (le_refl _)
+    obtain ⟨c, ⟨hc_pos, hc⟩⟩ := hn₀
+    refine ⟨c, ⟨hc_pos, ?_⟩⟩
+    rw [eventually_atTop]
+    exact ⟨n₀, hc⟩
```
@ocfnash I'm not sure how you'd get a contradiction in your first case. Isn't the point that norms can be negative if you only have `Norm`?

ADedecker commented:
Actually I have an even better suggestion: you should add 
```lean
lemma exists_eventually_atTop {α : Type _} [SemilatticeSup α] [Nonempty α] {r : α → β → Prop} :
    (∃ b, ∀ᶠ a in atTop, r a b) ↔ ∀ᶠ a₀ in atTop, ∃ b, ∀ a ≥ a₀, r a b := by
  simp_rw [eventually_atTop, ← exists_swap (α := α)]
  refine exists_congr fun a ↦ .symm <| forall_ge_iff <| Monotone.exists fun _ _ _ hb H n hn ↦
    H n (hb.trans hn)
```

Then, your proofs become:
```lean
theorem isBigO_atTop_iff_dep_const {α : Type _} [SemilatticeSup α] [Nonempty α]
    {f : α → E} {g : α → F} : f =O[atTop] g ↔ ∀ᶠ n₀ in atTop, ∃ c, ∀ n ≥ n₀, ‖f n‖ ≤ c * ‖g n‖ := by
  rw [isBigO_iff, exists_eventually_atTop]

theorem isBigO_atTop_iff_dep_const' {α : Type _}
    [SemilatticeSup α] [Nonempty α] {f : α → G} {g : α → G'} :
    f =O[atTop] g ↔ ∀ᶠ n₀ in atTop, ∃ c > 0, ∀ n ≥ n₀, c * ‖f n‖ ≤ ‖g n‖ := by
  simp_rw [isBigO_iff'', ← exists_prop, Subtype.exists', exists_eventually_atTop]
```

dupuisf commented:
Thanks a lot for the suggestions, Anatole, this is indeed much nicer! I have now implemented your last version, and changed the lemma names to what Oliver suggested.

ADedecker reviewed (approved):
@ocfnash I don't want to pull the trigger since I made quite heavy suggestions, would you mind having a final look?

ocfnash commented:
Thank you both, this is great!

bors merge
